### ISSUE TITLE
FR: Option to have workspaces independently from (multiple) displays / monitors

### ISSUE THREAD

**Issue by mrvantage** [pop-os/cosmic-comp#697]
Currently there are two options "Workspaces span multiple displays" and "Displays have separate workspaces". The span option really locks the applications on the first display to the the applications on the second, while the separate option should make it more dynamic to mix and match what workspaces are used together from the separate displays.

However the separate option provides a "deck" of workspaces per display, that are pretty much tied to the display. It is pretty hard to move a workspace from one display to another. Yes there are some key bindings to move a workspace to the stack of another display, but it would take multiple key combinations to get a specific workspace from another display into focus on the active display.

So, what I'm missing is an option to really mix and match the workspaces between displays, to be able to call any workspace to be displayed on any of the displays. For example Qtile has this feature.

Basically it comes down to this:

- There are 1 to n workspaces
- You can call any of the workspaces to the display in focus, by using a key combo (e.g. Super+[number]), or by clicking in a workplace selector
- If you call a workspace that is currently active on another display, the workspaces are switched between the displays.

I have no idea how hard it would be to implement this third mode of multi monitor behavior, or if it's even remotely possible. But this is a feature I'm really missing coming from Qtile.

**Comment by wolandark** [comment]
> to move a workspace to the stack of another display

That is not a thing in any WM/DE afaik. You don't move workspaces, you move the programs in them.

The fact that cosmic is providing _display independent workspaces_ is a huge leap in linux desktop development. The absence of this feature is why most developers use i3wm or other WMs.

Basically, as far as I know, the workspaces are numbered in order. So display 1 has workspaces 1 to 5 and display 2 has the rest of them, 6 to 10. To shuffle this order around should cause a lot of confusion and is probably pretty hard to implement in a solid way.

**Comment by cobbles** [comment]
I would like to add that being able to assign a workspace to a specific output like you can with swaywm would be a great improvement. Its a common workflow for tiling window managers to be able to "snap to" a workspace with a hotkey as @mrvantage describes. Keep up the great work and congrats on the alpha release 🙂

**Comment by n-kremeris** [comment]
> > to move a workspace to the stack of another display
> 
> That is not a thing in any WM/DE afaik. You don't move workspaces, you move the programs in them.
> 
> The fact that cosmic is providing _display independent workspaces_ is a huge leap in linux desktop development. The absence of this feature is why most developers use i3wm or other WMs.
> 
> Basically, as far as I know, the workspaces are numbered in order. So display 1 has workspaces 1 to 5 and display 2 has the rest of them, 6 to 10. To shuffle this order around should cause a lot of confusion and is probably pretty hard to implement in a solid way.

This is very much a thing in I3 at minimum, and likely a few other minimalistic WM's.

**Comment by TS-CUBED** [comment]
> > to move a workspace to the stack of another display
> 
> That is not a thing in any WM/DE afaik. You don't move workspaces, you move the programs in them.
> 
> The fact that cosmic is providing _display independent workspaces_ is a huge leap in linux desktop development. The absence of this feature is why most developers use i3wm or other WMs.
> 
> Basically, as far as I know, the workspaces are numbered in order. So display 1 has workspaces 1 to 5 and display 2 has the rest of them, 6 to 10. To shuffle this order around should cause a lot of confusion and is probably pretty hard to implement in a solid way.

As @n-kremeris said. This is very much a thing in i3. I used that all the time in teaching: set up all the required windows on the main laptop screen and with a single keyboard combo send that whole workspace to the other display (projector).

**Comment by mrvantage** [comment]
[This](https://github.com/pop-os/cosmic-epoch/issues/231) issue on cosmic-epoch seems to request the same thing

**Comment by bennidhamma** [comment]
Could this kind of thing be composed in a script, like i3-msg works?

**Comment by tlmoore-UniNA** [comment]
I also support global workspaces that can be assigned per monitor a la Sway WM/i3. Recently moved over to Cosmic from Sway WM and this the only stumbling block I've come across -- i.e. being able to assign all odd-numbered workspaces to a specific monitor and all even-numbered workspaces to a different monitor.
